clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -819,234 +819,47 @@ private ByteBuffer decompressRecord(CompressionStrategy compressionStrategy, Byt
+    if (handleCallbackForEmptyKeySet(keys, callback)) {
+      // empty key set
+      return;
     }
-
-    @Override
-    public void onCompletion(Optional<VeniceClientException> exception) {
-      // Only complete it when all the futures are done.
-      CompletableFuture.allOf(deserializationFutures.toArray(new CompletableFuture[deserializationFutures.size()]))
-          .whenComplete((voidP, throwable) -> {
-            Optional<Exception> completedException = Optional.empty();
-            if (exception.isPresent()) {
-              // Exception thrown by transporting layer
-              completedException = Optional.of(exception.get());
-            } else if (streamingFooterRecord.isPresent()) {
-              // Exception thrown by Venice backend
-              completedException = Optional.of(
-                  new VeniceClientHttpException(
-                      new String(ByteUtils.extractByteArray(streamingFooterRecord.get().detail)),
-                      streamingFooterRecord.get().status));
-            } else if (throwable != null) {
-              if (throwable instanceof Exception) {
-                completedException = Optional.of((Exception) throwable);
-              } else {
-                completedException = Optional.of(new Exception(throwable));
-              }
-            } else {
-              // Everything is good
-              if (!isStreamingResponse) {
-                /**
-                 * For regular response, the non-existing keys won't be present in the response,
-                 * so we need to manually collect the non-existing keys and trigger customer's callback.
-                 */
-                for (int i = 0; i < keyList.size(); ++i) {
-                  if (!receivedKeySet.get(i)) {
-                    callback.onRecordReceived(keyList.get(i), null);
-                    trackingStreamingCallback.ifPresent(t -> t.onRecordDeserialized());
-                    /**
-                     * To match the streaming response, we will mark the non-existing keys as received as well.
-                     */
-                    receivedKeySet.set(i);
-                  }
-                }
-              }
-            }
-            callback.onCompletion(completedException);
-            final Optional<Exception> finalCompletedException = completedException;
-            trackingStreamingCallback.ifPresent(
-                t -> t.onDeserializationCompletion(finalCompletedException, successfulKeyCnt.get(), duplicateEntryCnt));
-            clientStats.ifPresent(
-                stats -> stats.recordResponseDeserializationTime(
-                    LatencyUtils.convertLatencyFromNSToMS(deserializationTimeInNS.sum())));
-          });
+    List<K> keyList = new ArrayList<>(keys);
+    TransportClientStreamingCallback decoder = new MultiGetRecordStreamDecoder<>(
+        keyList,
+        callback,
+        callback,
```

Comment:
It was necessary because TrackingStreamingCallback didn't support callback composition, and not passing caller provided callback would disable metric collection at the decoder level.

Refactored the code in the latest commit to avoid passing callback twice. Thanks for the feedback!